iphone/cli/hooks/frameworks.js

```diff
@@ -0,0 +1,841 @@
+/**
+ * iOS build hook that scans for available frameworks from modules and in the
+ * project folder and then configures them in the Xcode project
+ */
+
+'use strict';
+
+const exec = require('child_process').exec;
+const fs = require('fs');
+const IncrementalFileTask = require('appc-tasks').IncrementalFileTask;
+const path = require('path');
+const wrench = require('wrench');
+
+const frameworkPattern = /([^/]+)\.framework$/;
+
+const FRAMEWORK_TYPE_STATIC = 'static';
+const FRAMEWORK_TYPE_DYNAMIC = 'dynamic';
+
+exports.cliVersion = '>=3.2.1';
+
+exports.init = function (logger, config, cli) {
+	let frameworkManager = new FrameworkManager(cli, config, logger);
+	frameworkManager.initialize();
+};
+
+/**
+ * Manages all available frameworks either from modules or the project itself
+ *
+ * Scans for available frameworks and inspects them to get the required info to
+ * properly integrate them into the Xcode project.
+ */
+class FrameworkManager {
+
+	/**
+	 * Constructs a new framework manager
+	 *
+	 * @param {Object} cli CLI instance
+	 * @param {Object} config Project configuration
+	 * @param {Object} logger Logger instance
+	 */
+	constructor(cli, config, logger) {
+		this._cli = cli;
+		this._config = config;
+		this._logger = logger;
+		this._builder = null;
+		this._frameworks = new Map();
+	}
+
+	/**
+	 * Initializes the framework manager by hooking into the required build steps
+	 */
+	initialize() {
+		this._cli.on('build.pre.compile', {
+			priority: 1200,
+			post: (builder, callback) => {
+				this._logger.trace('Starting third-party framework detection');
+				this._builder = builder;
+				this.detectFrameworks().then(callback, callback);
+			}
+		});
+
+		this._cli.on('build.ios.xcodeproject', {
+			pre: this.addFrameworksToXcodeProject.bind(this)
+		});
+	}
+
+	/**
+	 * Detects all available frameworks
+	 *
+	 * @return {Promise}
+	 */
+	detectFrameworks() {
+		return this.findFrameworkPaths().then((frameworkPaths) => {
```

Comment:
Generally, I recommend putting the `.then()` on the next line. This makes it a little more readable, especially if there's a `.catch()`.